### ISSUE TITLE
Utilization on Windows 10

### ISSUE THREAD

**Issue by pawlobanano** [furkilic/go-wrapper#4]
Hi,

it is a great idea for the wrapper. I'm already so used to gradlew - so naturally, I found your repository for my Golang learning journey.

I'm testing your solution on Windows 10. It works but in a strange way.

My example (terminal - Git bash from inside of Visual Studio Code):
MINGW64 /c/prj/go-tutorial-1/create-go-module/hello
$ ../gow.cmd run hello.go
find: ‘/i’: No such file or directory
find: ‘x86’: No such file or directory
go: cannot find GOROOT directory: "C:\prj\go-tutorial-1\create-go-module\.go\wrapper\go"

but Go sources are actually downloaded and unzipped fine.

Then for the run command I have to use 'gow' (bash version) because on 'gow.cmd' I got:
MINGW64 /c/prj/go-tutorial-1/create-go-module/hello
$ ../gow.cmd run hello.go
go: cannot find GOROOT directory: "C:\prj\go-tutorial-1\create-go-module\.go\wrapper\go"

So the result is:
MINGW64 /c/prj/go-tutorial-1/create-go-module/hello
$ ../gow run hello.go
greetings: empty name
exit status 1

And another thing - I insert the line:
distributionUrl=https://dl.google.com/go/go1.13.10.linux-amd64.tar.gz
into the .go/wrapper/go-wrapper.properties file but the downloaded sources are always in the latest version.

Seems like the project is not much active but maybe it is not that much effort to fix those? If you'd like, I'm ready to help to test it.

**Comment by furkilic** [comment]
Hi @pawlobanano 

Thanks for your report.
Indeed, I started writing `go-wrapper` during the best time possible (thx covid...) so the project is not that active.

But as you mentionned the issues should be quite easy to fix. 
The project need an update anyway.

I'll try to that by EOW. I'll let a message to let you know when it will be testable.

Thanks for your help and do not hesitate to look into https://github.com/furkilic/go-boot-config if you are familiar with spring-boot

Best regards

**Comment by furkilic** [comment]
Hi @pawlobanano 

Some changes have been push to ease the work on windows you can test the new scripts on the branch https://github.com/furkilic/go-wrapper/tree/misc/upgrade-project 

I'll merge the branch once you validate your tests.

I also added an FAQ in the readme  to answer some of the questions you asked yourself

## FAQ

#### Should i use `gow` or `gow.cmd`

Only on Windows and the `Command Prompt` or `PowerShell` use `./gow.cmd`

For all other options (_Windows with `GitBash` or `Cygwin`, Linux, MacOs.._) use `./gow`

#### Nothing is happening when I put an older `distributionUrl` in `.go/wrapper/go-wrapper.properties`

First of all ensure that  the `distributionUrl` matches the OS your are working on.

Ensure that there is not an already avalaible `GOROOT` in your Environment Variables

In case there is a previous Installation with `gow`, manually delete `.go/wrapper/tmp` and `.go/wrapper/go`



Thanks again 

Best regards

**Comment by pawlobanano** [comment]
Hi @furkilic
tl;dr - great job - it worked flawlessly - first try!:)

My quick tests
```
PBANACH@LCE43761 MINGW64 /c/prj/go-tutorial-1/create-go-module/hello
$ ../gow version
go version go1.13.10 windows/amd64

PBANACH@LCE43761 MINGW64 /c/prj/go-tutorial-1/create-go-module/hello
$ ../gow run hello.go 
map[Darrin:Great to see you, Darrin! Gladys:Great to see you, Gladys! Samantha:Hi, Samantha. Welcome!]
```
or cmd/powershell
```
C:\prj\go-tutorial-1\create-go-module\hello>..\gow.cmd version
go version go1.13.10 windows/amd64

C:\prj\go-tutorial-1\create-go-module\hello>..\gow.cmd run hello.go
map[Darrin:Hail, Darrin! Well met! Gladys:Hail, Gladys! Well met! Samantha:Hi, Samantha. Welcome!]
```

Go is a new language to me. I need to get used to the usage of mods and imports. As in my tutorial example, I have a go.mod file in the hello module depending on greeting module like that
```
module example.com/hello

go 1.16

replace example.com/greetings => ../greetings

require example.com/greetings v0.0.0-00010101000000-000000000000
```
this requires run of the gow/gow.cmd from inside of hello module directory - I mean

```
PBANACH@LCE43761 MINGW64 /c/prj/go-tutorial-1/create-go-module/hello
$ ../gow run hello.go 
```

I understand that it can't be run from the root directory
```
PBANACH@LCE43761 MINGW64 /c/prj/go-tutorial-1/create-go-module
$ ./gow run hello/hello.go
hello\hello.go:7:2: cannot find package "example.com/greetings" in any of:
        C:\prj\go-tutorial-1\create-go-module\.go\wrapper\go\src\example.com\greetings (from $GOROOT)        
        C:\Users\pbanach\go\src\example.com\greetings (from $GOPATH)
```
Same results with standard go approach - run by different go version as an alias
```
PBANACH@LCE43761 MINGW64 /c/prj/go-tutorial-1/create-go-module
$ go1.10.7.exe run hello/hello.go
hello\hello.go:7:2: cannot find package "example.com/greetings" in any of:
        C:\Users\pbanach\sdk\go1.10.7\src\example.com\greetings (from $GOROOT)
        C:\Users\pbanach\go\src\example.com\greetings (from $GOPATH)
```
Am I right assuming that this is normal Go flow, or could you explain this behaviour? Maybe _replace example.com/greetings => ../greetings_ can be defined differently?

And I have a question regarding the GOROOT env variable. Should it be set (as a variable in go-wrapper.properties) to existing go installation directory or just left unset? 
I mean it shouldn't be available in the terminal by _echo $GOROOT_ or _echo %GOROOT%_ Do I understand it right? 

Regards,
Paweł

**Comment by furkilic** [comment]
Thanks @pawlobanano 

For this specific example : `../gow run hello.go`  is the correct usage.
Note that your example is specific and based on a tutorial that is here to ease the starting for golang development.
Usually in GO there is no multi-module projects (similar to what you can find in java with maven or gradle)
It's usually one module per project with potentially multiple directories to organize the code (you can see more in https://github.com/golang-standards/project-layout). 
So there is a single `go.mod` at the root of the project, that's why normally you should not use gow outside of the root of the project.
To be compliant with your example, you'll need to have one `gow` inside `greetings` and also one inside `hello` as it is two different projects.
```
tree .
├───greetings
│          ├───go.mod
│          ├───gow
│          └───greetings.go
└───hello
       ├───go.mod
       ├───gow
       └───hello.go
```


Regarding `GOROOT` : 
The variable targets  the directory of your go installation (be carefull to not add the `bin`, it should really point to the root of the install folder, for ex : `GOROOT=C:/install/path/of/go`  )
If there is already an existing installation, the installer normally should have created the variable in your `User Environment Variables`  if not add it yourself. Even without using `gow`, you should do it, as this variable is used by `go` itself to retrieve internal libraries and tooling. 
`GOROOT` environment variable is a standard of `go` not `gow` so you don't need to add it in go-wrapper.properties
What `gow`  does behing the hood is, when `GOROOT` does not exists,  downloading the go install into `.go/wrapper/go` and pointing `GOROOT` to it

To summarize, if you already have an external installation of go, create the `User Environment Variable` `GOROOT` that points to it, so `echo $GOROOT` or `echo %GOROOT%` **should** return the path in the terminal. Once done, when using `gow`, it will automatically retrieve the external installation.

We went a bit out of scope of `gow` but anyway hope it helps.

I'll merge the PR.

**Comment by pawlobanano** [comment]
Yes, definitely a bit offtopic :)
Many thanks for the explanation. 

I see _GOROOT_ and _GOPATH_ in the output of  _go env_ command. 
And in windows environment variables, for the logged user, I have go locations mentioned in GOPATH and Path variables. 
And in system environment variables, another go locations assigned to the Path variable. 
A bit confusing but I will catch it soon.

With _gow env_ executed form the tutorial directory, I see the output:
```
set GOPATH=C:\Users\pbanach\go
set GOROOT=C:\prj\go-tutorial-1\create-go-module\.go\wrapper\go
set GOTOOLDIR=C:\prj\go-tutorial-1\create-go-module\.go\wrapper\go\pkg\tool\windows_amd64
```
and now I think I get it.

Anyway, I'm really glad _gow_ works. With it, it is now possible to completely avoid the separate system-wide installation of go which is a valuable feature.
